src/transformers/models/rwkv5/tokenization_rwkv5.py

```diff
@@ -0,0 +1,545 @@
+
+VOCAB_FILES_NAMES = {
+    "vocab_file": "rwkv_vocab_v20230424.txt",
+}
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "RWKV/rwkv-5-world-169m": "https://huggingface.co/RWKV/rwkv-5-world-169m/blob/main/rwkv_vocab_v20230424.txt",
+    },
+}
+
+
+class TRIE:
+    __slots__ = tuple("ch,to,values,front".split(","))
+    to: list
+    values: set
+
+    def __init__(self, front=None, ch=None):
+        self.ch = ch
+        self.to = [None for ch in range(256)]
+        self.values = set()
+        self.front = front
+
+    def __repr__(self):
+        fr = self
+        ret = []
+        while fr is not None:
+            if fr.ch is not None:
+                ret.append(fr.ch)
+            fr = fr.front
+        return "<TRIE %s %s>" % (ret[::-1], self.values)
+
+    def add(self, key: bytes, idx: int = 0, val=None):
+        if idx == len(key):
+            if val is None:
+                val = key
+            self.values.add(val)
+            return self
+        ch = key[idx]
+        if self.to[ch] is None:
+            self.to[ch] = TRIE(front=self, ch=ch)
+        return self.to[ch].add(key, idx=idx + 1, val=val)
+
+    def find_longest(self, key: bytes, idx: int = 0):
+        u: TRIE = self
+        ch: int = key[idx]
+
+        while u.to[ch] is not None:
+            u = u.to[ch]
+            idx += 1
+            if u.values:
+                ret = idx, u, u.values
+            if idx == len(key):
+                break
+            ch = key[idx]
+        return ret
+
+
+class RWKVWorldTokenizer(PreTrainedTokenizer):
+    vocab_files_names = VOCAB_FILES_NAMES
+    model_input_names = ["input_ids", "attention_mask"]
+
+    def __init__(self, vocab_file, errors="replace", pad_token="0", **kwargs):
+        self.add_bos_token = False
+        self.encoder = {}
+        sorted = []  # must be already sorted
+        with open(vocab_file, "r", encoding="utf-8") as f:
+            lines = f.readlines()
+        for l in lines:
+            idx = int(l[: l.index(" ")])
+            x = eval(l[l.index(" ") : l.rindex(" ")])
+            x = x.encode("utf-8") if isinstance(x, str) else x
+            assert isinstance(x, bytes)
+            assert len(x) == int(l[l.rindex(" ") :])
+            sorted += [x]
+            self.encoder[idx] = x
+
+        self.decoder = {}
+        for k, v in self.encoder.items():
+            self.decoder[v] = int(k)
+
+        self.trie = TRIE()
+        for t, i in self.decoder.items():
+            _ = self.trie.add(t, val=(t, i))
+        self.errors = errors  # how to handle errors in decoding
+        self.cache = {}
+        self.first_max_length = 0
+        super().__init__(
+            errors=errors,
+            **kwargs,
+        )
+
+    @property
+    def eos_token_id(self) -> Optional[int]:
+        return 0
+
+    @property
+    def eot_token_id(self) -> Optional[int]:
+        return 0
+
+    @property
+    def pad_token_id(self) -> Optional[int]:
+        return 0
+
```

Comment:
these should not be needed and should rely on the special tokens for eos and pad

Reply:
In RWKV tokenizer, the token id for bos, eos, and pad are all 0. This is not implemented by adding special tokens, but rather by directly rewriting the relevant attributes, because eos_token_id and pad_token_id are already being used by default in the decoding and batch encoding func of this tokenizer.

Reply:
Alright, this can be done with adding the content of the first token and init with eos_token = pad_token (the super class will have this information) but can be solved afterwards no worries it's minimal !